lib/consensus/raft/raft.go

```diff
@@ -0,0 +1,700 @@
+var errWaitingForSelf = errors.New("waiting for ourselves to depart")
+
+// RaftMaxSnapshots indicates how many snapshots to keep in the consensus data
+// folder.
+// TODO: Maybe include this in Config. Not sure how useful it is to touch
+// this anyways.
+var RaftMaxSnapshots = 5
+
+// RaftLogCacheSize is the maximum number of logs to cache in-memory.
+// This is used to reduce disk I/O for the recently committed entries.
+var RaftLogCacheSize = 512
+
+// How long we wait for updates during shutdown before snapshotting
+var waitForUpdatesShutdownTimeout = 5 * time.Second
+var waitForUpdatesInterval = 400 * time.Millisecond
+
+// How many times to retry snapshotting when shutting down
+var maxShutdownSnapshotRetries = 5
+
+// raftWrapper wraps the hraft.Raft object and related things like the
+// different stores used or the hraft.Configuration.
+// Its methods provide functionality for working with Raft.
+type raftWrapper struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	raft          *hraft.Raft
+	config        *config.ClusterRaftConfig
+	host          host.Host
+	serverConfig  hraft.Configuration
+	transport     *hraft.NetworkTransport
+	snapshotStore hraft.SnapshotStore
+	logStore      hraft.LogStore
+	stableStore   hraft.StableStore
+	//boltdb        *raftboltdb.BoltStore
+	staging bool
+}
+
+// newRaftWrapper creates a Raft instance and initializes
+// everything leaving it ready to use. Note, that Bootstrap() should be called
+// to make sure the raft instance is usable.
+func newRaftWrapper(
+	host host.Host,
+	cfg *config.ClusterRaftConfig,
+	fsm hraft.FSM,
+	staging bool,
+) (*raftWrapper, error) {
+
+	raftW := &raftWrapper{}
+	raftW.config = cfg
+	raftW.host = host
+	raftW.staging = staging
+	// Set correct LocalID
+	cfg.RaftConfig.LocalID = hraft.ServerID(peer.Encode(host.ID()))
+
+	//df := cfg.GetDataFolder()
+	//err := makeDataFolder(df)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	raftW.makeServerConfig()
+
+	err := raftW.makeTransport()
+	if err != nil {
+		return nil, err
+	}
+
+	err = raftW.makeStores()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debug("creating Raft")
+	raftW.raft, err = hraft.NewRaft(
+		cfg.RaftConfig,
+		fsm,
+		raftW.logStore,
+		raftW.stableStore,
+		raftW.snapshotStore,
+		raftW.transport,
+	)
+	if err != nil {
+		logger.Error("initializing raft: ", err)
+		return nil, err
+	}
+
+	raftW.ctx, raftW.cancel = context.WithCancel(context.Background())
+	//go raftW.observePeers()
+
+	return raftW, nil
+}
+
+// makeDataFolder creates the folder that is meant to store Raft data. Ensures
+// we always set 0700 mode.
+func makeDataFolder(folder string) error {
+	return os.MkdirAll(folder, 0700)
+}
+
+func (rw *raftWrapper) makeTransport() (err error) {
+	logger.Debug("creating libp2p Raft transport")
+	rw.transport, err = p2praft.NewLibp2pTransport(
+		rw.host,
+		rw.config.NetworkTimeout,
+	)
+	return err
+}
+
+func (rw *raftWrapper) makeStores() error {
```

Comment:
TBD if we want to use persistent stores here.

Reply:
Using persistent stores here